mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2010-2016. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.extensions.mongo.eventsourcing.eventstore;
+
+import com.mongodb.MongoClientSettings;
+import com.mongodb.ServerAddress;
+import com.mongodb.WriteConcern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * <p>
+ * Factory class used to create a {@code MongoClientSettings} instance. The instance makes use of the defaults as provided
+ * by the MongoOptions class. The moment you set a valid value, that value is used to create the options object.
+ * </p>
+ *
+ * Note: WriteConcern enums were changed in MongoDb driver 4.x.
+ * Depending on the number of addresses provided, the factory defaults to either {@link WriteConcern#W2} when
+ * more than one address is provided, or {@link WriteConcern#JOURNALED} when only one server is available. The idea of
+ * these defaults is that data must be able to survive a (not too heavy) crash without loss of data. We wouldn't want to
+ * publish untraceable events, would we...
+ *
+ * @author Jettro Coenradie
+ * @since 2.0 (in incubator since 0.7)
+ */
+public class MongoSettingsFactory {
+
+    private static final Logger logger = LoggerFactory.getLogger(MongoSettingsFactory.class);
+
+    private final MongoClientSettings defaults;
+    private List<ServerAddress> mongoAddresses = Collections.emptyList();
+    private WriteConcern writeConcern;
+    private int connectionsPerHost;
+    private int connectionTimeout;
+    private long maxWaitTime;
+    private int socketTimeOut;
```

Comment:
I still think we need to do something about having both the `connectionTimeout` and the `socketTimeOut`. 
As it currently stands, you're only using the `socketTimeOut` in the `MongoClientSettings`.

On top of that, we only set the `SocketSettings.connectTimeout` field, whilst previously we set a connection timeout _and_ socket timeout field. I do agree that Mongo's documentation on the matter is rather vague, so it'll be hard to deduce (without accessing their old implementation) what the original intent was of both fields.

As such I'd like to suggest the following:

1. Rename `connectionTimeout` to `socketConnectTimeout`. Adjust getter/setter and javadoc accordingly.
2. Rename `socketTimeOut` to `socketReadTimeout`. Adjust getter/setter and javadoc accordingly.
3. Use the `socketConnectTimeout` for `SocketBuilder.connectTimeout` method in the `createMongoClientSettings` function.
4. Use the `socketReadTimeout` for `SocketBuilder.readTimeout` method in the `createMongoClientSettings` function.
5. Adjust the debug logging in the `createMongoClientSettings` function to correctly reference all this information.

What's your thought about this @bjornharvold? Does this freedom of configuration make sense for your use case?

At the very least I hope to make things clear from this `MongoSettingsFactory` from here onward.

Reply:
This works.